src/library/scala/collection/SeqLike.scala

```diff
@@ -504,15 +504,20 @@ trait SeqLike[+A, +Repr] extends Any with IterableLike[A, Repr] with GenSeqLike[
    *  @return  A new $coll which contains the first occurrence of every element of this $coll.
    */
   def distinct: Repr = {
-    val b = newBuilder
-    val seen = mutable.HashSet[A]()
-    for (x <- this) {
-      if (!seen(x)) {
-        b += x
-        seen += x
+    val size = this.sizeHintIfCheap
+    if (size == 0 || size == 1 || isEmpty) repr
+    else {
+      val b = newBuilder
+      val seen = new mutable.HashSet[A]()
+      var current = thisCollection
+      var different = false
+      while (!current.isEmpty) {
+        val next = current.head
+        if (seen.add(next)) b += next else different = true
+        current = current.tail
```

Comment:
`.tail` is fast for `List`s but isn't guaranteed to be fast in general; I think this `while` optimization makes more sense as an override in `List` rather than here.

Reply:
Possibly it would be interesting to benchmark. Do you have any example of collection where you think it will be slow?
Also if that will be the case it might make sense to test while+ iterator vs for.

Reply:
In the extreme case, something like `ArrayBuffer` where `drop` needs to copy the entire collection.

Even something like `Vector` may need to do work to rebalance itself in order to remove an element. An `Iterator` is very likely to be better.

Reply:
We should add some non list test data and some mixed data types. Prove the case for non monomorohic dispatches. Don't think it will change the result by mixing. I have some ideas on the data set testing that we discussed previously.

Reply:
We should add some non list test data and some mixed data types. Prove the case for non monomorohic dispatches. Don't think it will change the result by mixing. I have some ideas on the data set testing that we discussed previously.

Reply:
I changed implementation to be different for `List` and for `SeqLike` and added test for `Vector`.
Still all results are much better than original.